<a href="https://colab.research.google.com/github/mikeninerbravog/POC_ML_AI-SPORTS/blob/master/POC_ML_AI_SPORTS.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [4]:
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestClassifier
import random

# 1. Simulação de Análise de Desempenho de Atletas
def analisar_desempenho():
    """
    Simula a análise de desempenho de atletas com base em três métricas:
    - Velocidade (km/h)
    - Resistência (%)
    - Precisão (%)

    O desempenho final é calculado como uma média ponderada dessas métricas.
    """
    np.random.seed(42)  # Define a semente para reprodutibilidade
    dados = {
        "atleta": [f"Atleta_{i}" for i in range(1, 21)],  # Cria 20 atletas fictícios
        "velocidade": np.random.uniform(20, 35, 20),  # Gera velocidades aleatórias entre 20 e 35 km/h
        "resistencia": np.random.uniform(60, 100, 20),  # Gera resistências entre 60% e 100%
        "precisao": np.random.uniform(50, 95, 20),  # Gera precisões entre 50% e 95%
    }
    df = pd.DataFrame(dados)  # Cria um DataFrame com os dados

    # Calcula a pontuação de desempenho ponderando as métricas
    df["pontuacao_desempenho"] = (df["velocidade"] * 0.3) + (df["resistencia"] * 0.4) + (df["precisao"] * 0.3)

    # Seleciona os 5 atletas com maior pontuação de desempenho
    melhores_atletas = df.nlargest(5, "pontuacao_desempenho")
    print("Top 5 Atletas Baseados no Desempenho:\n", melhores_atletas)
    return df

# 2. Previsão de Lesões com IA
def previsao_lesoes():
    """
    Simula um sistema de previsão de lesões usando IA.
    - Treina um modelo RandomForestClassifier para identificar risco de lesão
    - Considera a intensidade do treino e a fadiga acumulada como fatores principais
    """
    np.random.seed(42)  # Define a semente para reprodutibilidade
    atletas = np.array(range(1, 101))  # Cria 100 jogadores fictícios
    intensidade_treino = np.random.uniform(1, 10, 100)  # Intensidade do treino (1 a 10)
    fadiga = np.random.uniform(1, 10, 100)  # Fadiga acumulada (1 a 10)

    # Define uma regra simples: soma de treino e fadiga acima de 15 indica lesão
    lesoes = (intensidade_treino + fadiga > 15).astype(int)

    modelo = RandomForestClassifier()  # Cria um modelo de floresta aleatória
    modelo.fit(np.column_stack((intensidade_treino, fadiga)), lesoes)  # Treina o modelo

    # Dados de novos atletas para prever risco de lesão
    novos_dados = np.array([[8, 9], [3, 2], [7, 6]])
    previsoes = modelo.predict(novos_dados)  # Faz previsões de lesão

    print("Predições de lesões para novos atletas:", previsoes)
    return previsoes

# 3. Chatbot Esportivo Simples
def chatbot_esportivo(pergunta):
    """
    Chatbot esportivo simples para responder dúvidas sobre performance e IA nos esportes.
    Esse chatbot pode ser expandido para um assistente autônomo e interativo.
    O resultado aqui é apenas um exemplo para a PoC.
    """
    respostas = {
        "Como a IA ajuda os atletas?": "A IA analisa dados de desempenho e sugere treinos personalizados para melhorar a performance.",
        "Como prevenir lesões com IA?": "Monitorando fadiga e intensidade dos treinos, a IA pode prever riscos e sugerir descanso adequado.",
        "O que é análise tática esportiva?": "É o uso de IA para estudar padrões de jogo e criar estratégias mais eficientes.",
    }
    return respostas.get(pergunta, "Desculpe, não entendi sua pergunta.")

# Executando as funções
analisar_desempenho()
previsao_lesoes()

# Exemplo de interação com o chatbot
pergunta = "Como a IA ajuda os atletas?"
print(f"Usuário: {pergunta}")
print(f"Chatbot: {chatbot_esportivo(pergunta)}")


Top 5 Atletas Baseados no Desempenho:
        atleta  velocidade  resistencia   precisao  pontuacao_desempenho
13  Atleta_14   23.185087    97.955421  90.267231             73.217864
15  Atleta_16   22.751068    92.335894  91.484341             71.204980
14  Atleta_15   22.727375    98.625281  76.905499             69.339975
10  Atleta_11   20.308767    84.301794  93.631308             67.902740
5    Atleta_6   22.339918    91.407038  79.813503             67.208842
Predições de lesões para novos atletas: [1 0 0]
Usuário: Como a IA ajuda os atletas?
Chatbot: A IA analisa dados de desempenho e sugere treinos personalizados para melhorar a performance.


### **1. Análise de Desempenho dos Atletas**
O código avalia o desempenho dos atletas com base em três métricas principais:
- **Velocidade** (*km/h*) – Influencia 30% da pontuação.
- **Resistência** (*% de capacidade física*) – Influencia 40% da pontuação.
- **Precisão** (*% de acertos em um esporte específico*) – Influencia 30% da pontuação.

A pontuação final é calculada como:
\[
\text{Pontuação} = (\text{Velocidade} \times 0.3) + (\text{Resistência} \times 0.4) + (\text{Precisão} \times 0.3)
\]

Os **cinco atletas com melhor desempenho** foram selecionados com base nessa métrica:

| Atleta        | Velocidade (km/h) | Resistência (%) | Precisão (%) | Pontuação |
|--------------|------------------|---------------|------------|------------|
| Atleta_14   | 23.18            | 97.96         | 90.27      | **73.22**  |
| Atleta_16   | 22.75            | 92.34         | 91.48      | **71.20**  |
| Atleta_15   | 22.72            | 98.63         | 76.91      | **69.34**  |
| Atleta_11   | 20.31            | 84.30         | 93.63      | **67.90**  |
| Atleta_6    | 22.34            | 91.41         | 79.81      | **67.21**  |

**Observações sobre os melhores atletas:**
- **Atleta_14** teve o **maior desempenho geral**, pois sua resistência e precisão foram bastante altas, compensando uma velocidade moderada.
- **Atleta_16** também apresentou grande equilíbrio entre resistência e precisão, garantindo um segundo lugar.
- **Atleta_15** teve a **maior resistência** da lista, o que ajudou sua pontuação final.
- **Atleta_11** se destacou pela **alta precisão**, compensando sua menor velocidade.
- **Atleta_6** teve um bom equilíbrio entre as métricas.

---

### **2. Predição de Lesões**
O modelo de **RandomForestClassifier** foi treinado para prever lesões com base em:
- **Intensidade do treino** (*escala de 1 a 10*).
- **Fadiga acumulada** (*escala de 1 a 10*).

A regra definida no código diz que:
- Se **Intensidade do treino + Fadiga > 15**, o atleta tem **alto risco de lesão** (*marcado como 1*).
- Caso contrário, o atleta tem **baixo risco de lesão** (*marcado como 0*).

Os resultados para três novos atletas foram:
- **Atleta 1**: **Predição: 1 (Alto risco de lesão)** → Possivelmente devido à alta carga de treino e fadiga.
- **Atleta 2**: **Predição: 0 (Baixo risco de lesão)** → Intensidade e fadiga baixas.
- **Atleta 3**: **Predição: 0 (Baixo risco de lesão)** → Possivelmente treinando em um nível seguro.

---

### **3. Resposta do Chatbot**
O chatbot foi consultado com a pergunta **"Como a IA ajuda os atletas?"**, e respondeu:
> *"A IA analisa dados de desempenho e sugere treinos personalizados para melhorar a performance."*

Isso mostra como a IA pode ser utilizada para otimizar o treinamento e evitar lesões por meio da análise de dados.

---

### **Resumindo**
- O modelo foi capaz de identificar os **melhores atletas** com base em métricas objetivas.
- A IA previu corretamente **riscos de lesão** com base na carga de treino e fadiga.
- O chatbot forneceu uma **resposta relevante** sobre como a IA pode ajudar no desempenho esportivo.